chia/full_node/mempool_manager.py

```diff
@@ -286,42 +285,94 @@ async def pre_validate_spendbundle(
         )
         return ret
 
-    async def add_spendbundle(
+    async def add_spend_bundle(
         self,
         new_spend: SpendBundle,
         npc_result: NPCResult,
         spend_name: bytes32,
-        program: Optional[SerializedProgram] = None,
     ) -> Tuple[Optional[uint64], MempoolInclusionStatus, Optional[Err]]:
         """
-        Tries to add spend bundle to the mempool
-        Returns the cost (if SUCCESS), the result (MempoolInclusion status), and an optional error
+        Validates and adds to mempool a new_spend with the given NPCResult, and spend_name, and the current mempool.
+        The mempool should be locked during this call (blockchain lock). If there are mempool conflicts, the conflicting
+        spends might be removed (if the new spend is a superset of the previous). Otherwise, the new spend might be
+        added to the potential pool.
+
+        Args:
+            new_spend: spend bundle to validate and add
+            npc_result: result of running the clvm transaction in a fake block
+            spend_name: hash of the spend bundle data, passed in as an optimization
+
+        Returns:
+            Optional[uint64]: cost of the entire transaction, None iff status is FAILED
+            MempoolInclusionStatus:  SUCCESS (should add to pool), FAILED (cannot add), and PENDING (can add later)
+            Optional[Err]: None iff status is FAILED
+        """
+
+        # Skip if already added
+        if spend_name in self.mempool.spends:
+            cost: Optional[uint64] = self.mempool.spends[spend_name].cost
+            assert cost is not None
+            return uint64(cost), MempoolInclusionStatus.SUCCESS, None
+
+        cost, status, err, item, remove_items = await self.validate_spend_bundle(new_spend, npc_result, spend_name)
```

Comment:
I have a feeling the return values can be simplified here. specifically, we just pass through `err` here. Do we really want to do the work below if the validation returns an error?
I suspect that perhaps there's some implied invariant between whether we get an error and the `status` this returns.

Reply:
OK i'll try that

Reply:
I can't think of anything obvious. I think because of the lack of (or awkward use of) variant types in python, we just put all possible return types in a tuple, even if some are mutually exclusive.